File: ios/RetenoUserAttributes.swift
//
//  RetenoUserAttributes.swift
//  RetenoSdk
//
//  Created by Valentyn Halkin on 14.12.2022.
//  Copyright © 2022 Facebook. All rights reserved.
//

import Foundation
import Reteno

struct RetenoUserAttributesFieldStruct: Codable {
    let key: String
    let value: String
    init(dictionary: [String: Any]) throws {
        self = try JSONDecoder().decode(
            RetenoUserAttributesFieldStruct.self,
            from: JSONSerialization.data(withJSONObject: dictionary)
        )
    }
}

struct RetenoUserAttributesAddressStruct: Codable {
    init(dictionary: [String: Any]) throws {
        self = try JSONDecoder().decode(
            RetenoUserAttributesAddressStruct.self,
            from: JSONSerialization.data(withJSONObject: dictionary)
        )
    }
    let region: String?
    let town: String?
    let address: String?
    let postcode: String?
}

struct RetenoUserAttributesStruct: Codable {
    init(dictionary: [String: Any]) throws {
        self = try JSONDecoder().decode(
            RetenoUserAttributesStruct.self,
            from: JSONSerialization.data(withJSONObject: dictionary)
        )
    }
    let phone: String?
    let email: String?
    let firstName: String?
    let lastName: String?
    let languageCode: String?
    let timeZone: String?
    let address: RetenoUserAttributesAddressStruct?
    let fields: [RetenoUserAttributesFieldStruct]?
}

struct RetenoUserStruct: Codable {
    init(dictionary: [String: Any]?) throws {
        self = try JSONDecoder().decode(RetenoUserStruct.self, from: JSONSerialization.data(withJSONObject: dictionary ?? [:]))
    }
    let userAttributes: RetenoUserAttributesStruct?
    let subscriptionKeys: [String]?
    let groupNamesInclude: [String]?
    let groupNamesExclude: [String]?
}

public struct RetenoSetUserAttributesPayload: Codable {
    init(userAttributes: UserAttributes?, subscriptionKeys: [String], groupNamesInclude: [String], groupNamesExclude: [String]) {
        self.userAttributes = userAttributes;
        self.subscriptionKeys = subscriptionKeys;
        self.groupNamesInclude = groupNamesInclude;
        self.groupNamesExclude = groupNamesExclude;
    }

    let userAttributes: UserAttributes?
    let subscriptionKeys: [String]
    let groupNamesInclude: [String]
    let groupNamesExclude: [String]
}

public struct RetenoUserAttributes {
    
    
    private init() {}
    
    private static func getStringOrNil(input: String?) -> String? {
        return (input ?? "").isEmpty ? nil : input!
    }
    
    private static func buildUserAttributes(userStruct: RetenoUserStruct) -> UserAttributes {
        let fields = userStruct.userAttributes?.fields?.map { field in
            UserCustomField(key: field.key, value: field.value)
        }
        
        return UserAttributes(
            phone: getStringOrNil(input: userStruct.userAttributes?.phone),
            email: getStringOrNil(input: userStruct.userAttributes?.email),
            firstName: getStringOrNil(input: userStruct.userAttributes?.firstName),
            lastName: getStringOrNil(input: userStruct.userAttributes?.lastName),
            languageCode: getStringOrNil(input: userStruct.userAttributes?.languageCode),
            timeZone: getStringOrNil(input: userStruct.userAttributes?.timeZone),
            address: userStruct.userAttributes?.address != nil ? Address(
                region: getStringOrNil(input: userStruct.userAttributes?.address?.region),
                town: getStringOrNil(input: userStruct.userAttributes?.address?.town),
                address: getStringOrNil(input: userStruct.userAttributes?.address?.address),
                postcode: getStringOrNil(input: userStruct.userAttributes?.address?.postcode)
            ) : nil,
            fields: fields ?? []
        );
    }
    
    public static func buildSetUserAttributesPayload(payload: NSDictionary) throws -> RetenoSetUserAttributesPayload {
        let user = payload["user"] as? [String: Any];
        do {
            let userStruct = try RetenoUserStruct(dictionary: user);
            let userAttributes = userStruct.userAttributes != nil ? buildUserAttributes(userStruct: userStruct) : nil;
            let requestPayload = RetenoSetUserAttributesPayload(
                userAttributes: userAttributes,
                subscriptionKeys: userStruct.subscriptionKeys ?? [],
                groupNamesInclude: userStruct.groupNamesInclude ?? [],
                groupNamesExclude: userStruct.groupNamesExclude ?? []
            )
            
            return requestPayload;
        } catch {
            throw error;
        }
    }
    
    
}

File: android/src/main/java/com/retenosdk/RetenoUserAttributes.java
package com.retenosdk;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableArray;
import com.reteno.core.domain.model.user.Address;
import com.reteno.core.domain.model.user.User;
import com.reteno.core.domain.model.user.UserAttributes;
import com.reteno.core.domain.model.user.UserCustomField;

import java.util.ArrayList;
import java.util.List;

public class RetenoUserAttributes {
  private static List<UserCustomField> buildUserCustomData(ReadableArray fields) {
    int countView = fields.size();
    if (countView == 0) return null;

    List<UserCustomField> list = new ArrayList<>();
    for (int i = 0; i < countView; i++) {
      ReadableMap field = fields.getMap(i);

      String key = null;
      String value = null;

      if (field.getString("key") != null) {
        key = field.getString("key");
      }
      if (field.getString("value") != null) {
        value = field.getString("value");
      }

      if (key != null) {
        list.add(new UserCustomField(key, value));
      }
    }
    return list;
  }

  private static List<String> buildStringArr(ReadableMap user, String key) {
    if (user == null) {
      return null;
    }
    ReadableArray payloadStringArr = user.getArray("subscriptionKeys");
    if (payloadStringArr == null) {
      return null;
    }

    int countView = payloadStringArr.size();
    if (countView == 0) return null;

    List<String> stringArr = new ArrayList<>();
    for (int i = 0; i < countView; i++) {
      String str = payloadStringArr.getString(i);
      stringArr.add(str);
    }
    return stringArr;
  }

  public static User buildUserFromPayload(ReadableMap payload) {
    ReadableMap payloadUser = null;
    ReadableMap payloadUserAttributes = null;

    String payloadPhone = null;
    String payloadEmail = null;
    String payloadFirstName = null;
    String payloadLastName = null;
    String payloadLanguageCode = null;
    String payloadTimeZone = null;

    ReadableMap payloadAddress = null;
    ReadableArray payloadFields = null;

    Address address = null;
    List<UserCustomField> fields = null;

    payloadUser = payload.getMap("user");

    if (payloadUser != null) {
      payloadUserAttributes = payloadUser.getMap("userAttributes");
      if (payloadUserAttributes != null) {
        payloadPhone = payloadUserAttributes.getString("phone");
        payloadEmail = payloadUserAttributes.getString("email");
        payloadFirstName = payloadUserAttributes.getString("firstName");
        payloadLastName = payloadUserAttributes.getString("lastName");
        payloadLanguageCode = payloadUserAttributes.getString("languageCode");
        payloadTimeZone = payloadUserAttributes.getString("timeZone");
        payloadAddress = payloadUserAttributes.getMap("address");
        payloadFields = payloadUserAttributes.getArray("fields");
      }
    }

    if (payloadAddress != null) {
      address = new Address(
        RetenoUtil.getStringOrNull(payloadAddress.getString("region")),
        RetenoUtil.getStringOrNull(payloadAddress.getString("town")),
        RetenoUtil.getStringOrNull(payloadAddress.getString("address")),
        RetenoUtil.getStringOrNull(payloadAddress.getString("postcode"))
      );
    }

    if (payloadFields != null) {
      fields = buildUserCustomData(payloadFields);
    }

    UserAttributes userAttributes = new UserAttributes(
      RetenoUtil.getStringOrNull(payloadPhone),
      RetenoUtil.getStringOrNull(payloadEmail),
      RetenoUtil.getStringOrNull(payloadFirstName),
      RetenoUtil.getStringOrNull(payloadLastName),
      RetenoUtil.getStringOrNull(payloadLanguageCode),
      RetenoUtil.getStringOrNull(payloadTimeZone),
      address,
      fields
    );

    List<String> subscriptionKeys = buildStringArr(payloadUser, "subscriptionKeys");
    List<String> groupNamesInclude = buildStringArr(payloadUser, "groupNamesInclude");
    List<String> groupNamesExclude = buildStringArr(payloadUser, "groupNamesExclude");

    return new User(userAttributes, subscriptionKeys, groupNamesInclude, groupNamesExclude);
  }
}
